File: app/src/main/java/org/ionchain/wallet/mvp/model/home/HomePageModel.java
package org.ionchain.wallet.mvp.model.home;

import android.util.Log;

import com.lzy.okgo.callback.StringCallback;
import com.lzy.okgo.model.HttpParams;
import com.lzy.okgo.model.Response;
import com.lzy.okgo.request.base.Request;

import org.ionchain.wallet.bean.DeviceBean;
import org.ionchain.wallet.bean.DeviceListBean;
import org.ionchain.wallet.bean.WalletBean;
import org.ionchain.wallet.mvp.callback.OnBindDeviceCallback;
import org.ionchain.wallet.mvp.callback.OnDeviceDetailCallback;
import org.ionchain.wallet.mvp.callback.OnDeviceListCallback;
import org.ionchain.wallet.mvp.callback.OnUnbindDeviceCallback;
import org.ionchain.wallet.utils.NetUtils;
import org.ionchain.wallet.utils.ToastUtil;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.ionchain.wallet.constant.ConstantUrl.DEVICES_BIND_POST;
import static org.ionchain.wallet.constant.ConstantUrl.DEVICES_GET;
import static org.ionchain.wallet.constant.ConstantUrl.DEVICES_UNBIND_POST;

/**
 * USER: binny
 * DATE: 2018/9/13
 * 描述: 首页数据的业务类
 */
public class HomePageModel implements IHomePageModel {
    private String TAG = "HomePageModel";
    private HttpParams params;

    public HomePageModel() {
        params = new HttpParams();
    }

    @Override
    public void getCurrentWalletDevicesList(WalletBean walletBean, final OnDeviceListCallback callback) {
        params.clear();
        params.put("eth_address", walletBean.getAddress());
        NetUtils.get(DEVICES_GET, params, new StringCallback() {

            @Override
            public void onStart(Request<String, ? extends Request> request) {
                super.onStart(request);
                callback.onLoadStart();
            }

            @Override
            public void onSuccess(Response<String> response) {
                String json = response.body();
                Log.i(TAG, "onCreateSuccess: " + json);
                DeviceListBean bean = NetUtils.gsonToBean(json, DeviceListBean.class);
                if (bean == null || bean.getData() == null) {
                    return;
                }
                callback.onDeviceListSuccess(bean.getData());

            }

            @Override
            public void onError(Response<String> response) {
                super.onError(response);
                callback.onDeviceListFailure(response.body());
            }

            @Override
            public void onFinish() {
                super.onFinish();
                callback.onLoadFinish();
            }
        }, callback);

    }

    @Override
    public void getAllWalletDeviceList(List<WalletBean> walletBeans, OnDeviceListCallback listData) {

    }

    @Override
    public void bindDeviceToWallet(String address, String cksn, final OnBindDeviceCallback callback) {
        params.clear();
        params.put("eth_address", address);
        params.put("cksn", cksn);
        NetUtils.post(DEVICES_BIND_POST, params, new StringCallback() {
            @Override
            public void onSuccess(Response<String> response) {
                if (response == null) {
                    ToastUtil.showToastLonger("绑定失败");
                    return;
                }

                String json = response.body();
                DeviceBean bindBean = NetUtils.gsonToBean(json, DeviceBean.class);
                if (bindBean == null || bindBean.getData() == null) {
                    callback.onBindFailure("绑定失败");
                    return;
                }
                if (bindBean.getSuccess() != 0) {
                    ToastUtil.showToastLonger(bindBean.getMessage());
                } else {
                    callback.onBindSuccess(bindBean.getData());
                }
            }

            @Override
            public void onStart(Request<String, ? extends Request> request) {
                super.onStart(request);
                callback.onLoadStart();
            }

            @Override
            public void onError(Response<String> response) {
                super.onError(response);
                callback.onBindFailure(response.body());
            }

            @Override
            public void onFinish() {
                super.onFinish();
                callback.onLoadFinish();
            }
        }, callback);

    }

    @Override
    public void unbindDeviceToWallet(String address, String cksn, final OnUnbindDeviceCallback callback) {
        Map<String, String> params = new HashMap<>();
        params.put("eth_address", address);
        params.put("cksn", cksn);
        JSONObject jsonObject = new JSONObject(params);
        NetUtils.post(DEVICES_UNBIND_POST, jsonObject, new StringCallback() {
            @Override
            public void onSuccess(Response<String> response) {
                if (response == null) {
                    ToastUtil.showToastLonger("绑定失败");
                    return;
                }

                String json = response.body();
                DeviceBean bindBean = NetUtils.gsonToBean(json, DeviceBean.class);
                if (bindBean == null || bindBean.getData() == null) {
                    callback.onUnbindFailure("解绑定失败");
                    return;
                }
                if (bindBean.getSuccess() != 0) {
                    ToastUtil.showToastLonger(bindBean.getMessage());
                } else {
                    callback.onUnbindSuccess(bindBean.getData());
                }
            }

            @Override
            public void onStart(Request<String, ? extends Request> request) {
                super.onStart(request);
                callback.onLoadStart();
            }

            @Override
            public void onError(Response<String> response) {
                super.onError(response);
                callback.onUnbindFailure(response.body());
            }

            @Override
            public void onFinish() {
                super.onFinish();
                callback.onLoadFinish();
            }
        }, callback);
    }

    @Override
    public void getDeviceDetail(String cksn, OnDeviceDetailCallback deviceBean) {

    }
}

File: app/src/main/java/org/ionchain/wallet/mvp/model/home/HomePageModel.kt
package org.ionchain.wallet.mvp.model.home

import com.lzy.okgo.callback.StringCallback
import com.lzy.okgo.model.HttpParams
import com.lzy.okgo.model.Response
import com.lzy.okgo.request.base.Request
import org.ionc.wallet.bean.WalletBean
import org.ionc.wallet.utils.Logger
import org.ionchain.wallet.App
import org.ionchain.wallet.R
import org.ionchain.wallet.bean.DeviceBean
import org.ionchain.wallet.bean.DeviceListBean
import org.ionchain.wallet.constant.ConstantErrorCode.*
import org.ionchain.wallet.constant.ConstantUrl.*
import org.ionchain.wallet.mvp.callback.OnBindDeviceCallback
import org.ionchain.wallet.mvp.callback.OnDeviceDetailCallback
import org.ionchain.wallet.mvp.callback.OnDeviceListCallback
import org.ionchain.wallet.mvp.callback.OnUnbindDeviceCallback
import org.ionchain.wallet.utils.NetUtils
import org.ionchain.wallet.utils.ToastUtil
import org.json.JSONObject
import java.util.*

/**
 * USER: binny
 * DATE: 2018/9/13
 * 描述: 首页数据的业务类
 */
class HomePageModel : IHomePageModel {
    private val TAG = "HomePageModel"
    private val params: HttpParams = HttpParams()

    override fun getCurrentWalletDevicesList(walletBean: WalletBean, callback: OnDeviceListCallback) {
        params.clear()
        params.put("eth_address", walletBean.address)
        NetUtils.get(DEVICES_GET, params, object : StringCallback() {

            override fun onStart(request: Request<String, out Request<*, *>>?) {
                super.onStart(request)
                callback.onLoadStart()
            }

            override fun onSuccess(response: Response<String>) {
                val json = response.body()
                Logger.j(TAG, "onCreateSuccess: $json")
                val bean = NetUtils.gsonToBean(json, DeviceListBean::class.java)
                if (bean?.data == null) {
                    callback.onLoadFinish()
                    return
                }
                callback.onDeviceListSuccess(bean.data!!)

            }

            override fun onError(response: Response<String>) {
                super.onError(response)
                if (response.body() == null) {
                    callback.onDeviceListFailure(App.mContext.resources.getString(R.string.device_service_error))
                } else {
                    callback.onDeviceListFailure(response.body())
                }
            }

            override fun onFinish() {
                super.onFinish()
                callback.onLoadFinish()
            }
        }, callback)

    }

    override fun getAllWalletDeviceList(walletBeans: List<WalletBean>, listData: OnDeviceListCallback) {

    }

    override fun bindDeviceToWallet(address: String, cksn: String, callback: OnBindDeviceCallback) {
        params.clear()
        params.put("eth_address", address)
        params.put("cksn", cksn)
        NetUtils.post(DEVICES_BIND_POST, params, object : StringCallback() {
            override fun onSuccess(response: Response<String>?) {
                if (response == null) {
                    ToastUtil.showToastLonger(ERROR_CODE_NET_WORK_ERROR_DEVICE_BIND_NULL)
                    return
                }

                val json = response.body()
                val bindBean = NetUtils.gsonToBean(json, DeviceBean::class.java)
                if (bindBean == null || bindBean.data == null) {
                    callback.onBindFailure(ERROR_CODE_NET_WORK_ERROR_DEVICE_BIND_NULL)
                    return
                }
                if (bindBean.success != 0) {
                    ToastUtil.showToastLonger(bindBean.message)
                } else {
                    callback.onBindSuccess(bindBean.data!!)
                }
            }

            override fun onStart(request: Request<String, out Request<*, *>>?) {
                super.onStart(request)
                callback.onLoadStart()
            }

            override fun onError(response: Response<String>) {
                super.onError(response)
                Logger.i(response.exception.message.orEmpty())
                callback.onBindFailure( ERROR_CODE_NET_WORK_ERROR_DEVICE_BIND)
            }

            override fun onFinish() {
                super.onFinish()
                callback.onLoadFinish()
            }
        }, callback)

    }

    override fun unbindDeviceToWallet(address: String, cksn: String, callback: OnUnbindDeviceCallback) {
        val params = HashMap<String, String>()
        params["eth_address"] = address
        params["cksn"] = cksn
        val jsonObject = JSONObject(params)
        NetUtils.post(DEVICES_UNBIND_POST, jsonObject, object : StringCallback() {
            override fun onSuccess(response: Response<String>?) {
                if (response == null) {
                    ToastUtil.showToastLonger(ERROR_CODE_NET_WORK_ERROR_DEVICE_UNBIND_NULL)
                    return
                }

                val json = response.body()
                val bindBean = NetUtils.gsonToBean(json, DeviceBean::class.java)
                if (bindBean?.data == null) {
                    callback.onUnbindFailure(ERROR_CODE_NET_WORK_ERROR_DEVICE_UNBIND_NULL)
                    return
                }
                if (bindBean.success != 0) {
                    ToastUtil.showToastLonger(bindBean.message)
                } else {
                    callback.onUnbindSuccess(bindBean.data!!)
                }
            }

            override fun onStart(request: Request<String, out Request<*, *>>?) {
                super.onStart(request)
                callback.onLoadStart()
            }

            override fun onError(response: Response<String>) {
                super.onError(response)
                callback.onUnbindFailure(ERROR_CODE_NET_WORK_ERROR_DEVICE_UNBIND)
            }

            override fun onFinish() {
                super.onFinish()
                callback.onLoadFinish()
            }
        }, callback)
    }

    override fun getDeviceDetail(cksn: String, deviceBean: OnDeviceDetailCallback) {

    }
}
